### ISSUE TITLE
`prefer-global-this` should allow `window.setTimeout()` and `setInterval()`

### ISSUE THREAD

**Issue by n0099** [sindresorhus/eslint-plugin-unicorn#2482]
### Description

as it's a common workaround: [1](https://stackoverflow.com/questions/45802988/typescript-use-correct-version-of-settimeout-node-vs-window), [2](https://stackoverflow.com/questions/51040703/what-return-type-should-be-used-for-settimeout-in-typescript), for `globalThis.setTimeout()` and `setInterval()` returning [`NodeJS.Timeout`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/367fcd5cc72baeab1c7fa927f774a36ec7685c69/types/node/timers.d.ts#L90) in [`@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/367fcd5cc72baeab1c7fa927f774a36ec7685c69/types/node/timers.d.ts#L159)

### Fail

```ts
const id: number = window.setTimeout(console.log, 100);
//                 ^^^^^^
// Prefer `globalThis` over `window`. eslint(unicorn/prefer-global-this)
```

### Pass

```ts
const id: number = window.setTimeout(console.log, 100);
```
```ts
const id: NodeJS.Timeout | number = globalThis.setTimeout(console.log, 100);
```

**Comment by sindresorhus** [comment]
I disagree. The correct way to handle it is:

```js
const id: ReturnType<typeof setTimeout> = window.setTimeout(console.log, 100);
```

**Comment by fregante** [comment]
I had this exact issue yesterday 😅 

I wish there was a way to instruct the user on how to proceed in cases like this. Should this specific scenario be handled differently? For example:

- disable auto-fix in TS files
- extend the message to specify how to specify the type
- mention it in the rule's readme

I also had issues regarding `globalThis` types after running this rule. "Nobody knows" how to make globals available as properties on `globalThis` (the solution seems to be `declare var someGlobal: Type`, see https://github.com/xojs/eslint-config-xo-typescript/issues/91)

**Comment by n0099** [comment]
In my [scenario](https://github.com/n0099/open-tbm/commit/7b9cc7f046e809cc0066e5a7240ba0b43457b684#diff-0960cd0f015635f9ed0b6506800c11e4f3f4dd7423c9e4b208d8f1183552b146R41-R34) I'm storing the id returned by `setTimeout()` for latter cancelation, if I have to make it compatible with both node and browser enviornment, the initial value of id should be conditionally set to `let id = import.meta.client ? 0 : import.meta.server ? new NodeJS.Timeout() : ...;` for preventing [`let` TDZ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz) and violation of eslint rule [`init-declarations`](https://eslint.org/docs/latest/rules/init-declarations)

**Comment by fregante** [comment]
You don't need any of that. Just call setTimeout normally. 

As for the let, the lint rule is right. This is correct:

```ts
let id: ReturnType<typeof setTimeout> | undefined;

**Comment by n0099** [comment]
> for preventing [`let` TDZ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)

`let id;` is a TDZ, and I have to `clearTimeout()` before setting a new one

**Comment by n0099** [comment]
> This is correct:

with `// eslint-disable-next-line init-declarations`

**Comment by fregante** [comment]
That rules isn't enabled in any recommended config: https://eslint.org/docs/latest/rules/init-declarations

As you said, it's completely at odds with `unicorn/no-useless-undefined` so pick one or the other.

**Comment by xsjcTony** [comment]
> I disagree. The correct way to handle it is:
> 
> ```js
> const id: ReturnType<typeof setTimeout> = window.setTimeout(console.log, 100);
> ```

It won't work. It will always default to NodeJS's version unless you explicitly specify `window.setXxx`
![image](https://github.com/user-attachments/assets/1171f17e-9371-45fa-961e-de2135bc8d92)
![image](https://github.com/user-attachments/assets/4dbb6d4b-3fcd-42c2-a08a-0a3621ba373b)

**Comment by fregante** [comment]
That was a typo, drop `window` as the rule suggests

**Comment by n0099** [comment]
`typeof setTimeout` is referring `typeof globalThis.setTimeout`, so it's not the same with your `window.setTimeout`

**Comment by xsjcTony** [comment]
I've tried all the different combinations, it will always default to Node.js's version🫠
eslint-disable for now I guess
![image](https://github.com/user-attachments/assets/9d34653e-09b3-4776-9dd6-d63527879af8)
![image](https://github.com/user-attachments/assets/8e22d278-2986-4bb4-b2ea-07596afe061b)
![image](https://github.com/user-attachments/assets/ae447814-c693-41b4-8cd3-36cac78af7bd)

**Comment by fregante** [comment]
> `typeof setTimeout` is referring `typeof globalThis.setTimeout`, so it's not the same with your `window.setTimeout`

It doesn't matter as long as you don't use any specific features from node's.

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAcgHYQCYFMEG4BQODGESAzvMKgFxwBK6MArlEgCoCeY6APDO+hAGZxidZqD70YAPjgBeISLEQJACkIkIAG3QA6DRADmAGjgBGAAxmAlLji24AentwAegH4gA

In short, use the `setTimeout` function as you've always done: without `window`

**Comment by xsjcTony** [comment]
Yes it will work if the only usage is for clearInterval. However my function here requires a `number` type.
![image](https://github.com/user-attachments/assets/ece3949b-81e8-4f50-a769-da051893cc7b)
Guess I need to change `number` to `ReturnType<typeof setTimeout>`, but that's just so weird🫠

**Comment by n0099** [comment]
> [...]it will always default to Node.js's version🫠[...]
> [...]
> ![image](https://private-user-images.githubusercontent.com/45784210/380035109-ae447814-c693-41b4-8cd3-36cac78af7bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4Mzc4NTksIm5iZiI6MTcyOTgzNzU1OSwicGF0aCI6Ii80NTc4NDIxMC8zODAwMzUxMDktYWU0NDc4MTQtYzY5My00MWI0LThjZDMtMzZjYWM3OGFmN2JkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDA2MjU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjMWNmYjY1ODNlYzE4ODI1NzM0ODA4NjA3ZTI4ODdiNDc3M2I0YTQzMzRjYzQ1MzVjZjQ3YzUwMDcxMWRiNGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2MXSR4mikssJHHcnME435x2D-aZE_jLZcGJim4JD9zs)

`window.setTimeout` is the one for browser enviorment.

**Comment by xsjcTony** [comment]
Honestly, I agree with adding this special case to the rule, as it's so classical and common that both versions will be used in a project.
The most straightforward way to differentiate is to use `window.setTimeout` for the browser and `globalThis.setTimeout` for the node.

**Comment by fregante** [comment]
> `window.setTimeout` is the one for browser enviorment.


That's completely irrelevant. This is a type-only issue, `window === globalThis` in real life.

I'm closing this for now because the discussion is going nowhere. This is a documentation issue/misunderstanding. I'll open a dedicated issue to improve the DX as described [earlier](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2482#issuecomment-2424797467)

**Comment by xsjcTony** [comment]
> > `window.setTimeout` is the one for browser enviorment.
> 
> That's completely irrelevant. This is a type-only issue, `window === globalThis` in real life.
> 
> I'm closing this for now because the discussion is going nowhere. This is a documentation issue/misunderstanding. I'll open a dedicated issue to improve the DX as described [earlier](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2482#issuecomment-2424797467)

Indeed, I ran in to `no-var` as well today.😂 Yeah let's do a separate issue for those things, I'll just eslint-disable for now.

@n0099 Yes I do understand Web API and Node.js API and that's exactly why I'm here. I still think using `window.setTimeout` is the most clear way, but anyways.

**Comment by n0099** [comment]
> That's completely irrelevant. This is a type-only issue, `window === globalThis` in real life.

I know that both `window.setTimeout` and `globalThis.setTimeout` will return `NodeJS.Timeout` on node env.
But only the typing for `globalThis.setTimeout` is defined in [`@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/367fcd5cc72baeab1c7fa927f774a36ec7685c69/types/node/timers.d.ts#L152), that's why sticking with `window.setTimeout` will use the one from [`lib.dom.d.ts`](https://github.com/microsoft/TypeScript/blob/df9d16503f6755dd071e4c591b9d21c39d03d95e/src/lib/dom.generated.d.ts#L26650) which comes with typescript itself and cannot get removed like `@types/node`.
